diff --git a/include/synchronizer.h b/include/synchronizer.h
index 326a39f..1c806e0 100644
--- a/include/synchronizer.h
+++ b/include/synchronizer.h
@@ -12,6 +12,11 @@ class Synchronizer_Common
 {
 protected:
     Synchronizer_Common() {}
+    ~Synchronizer_Common() {
+        begin_atomic();
+        wakeup_all();
+        end_atomic();
+    }
 
     // Atomic operations
     bool tsl(volatile bool & lock) { return CPU::tsl(lock); }
@@ -22,9 +27,56 @@ protected:
     void begin_atomic() { Thread::lock(); }
     void end_atomic() { Thread::unlock(); }
 
-    void sleep() { Thread::yield(); } // implicit unlock()
-    void wakeup() { end_atomic(); }
-    void wakeup_all() { end_atomic(); }
+    void sleep() { 
+        begin_atomic();
+
+        Thread* previous = Thread::running();
+        previous->_state = Thread::WAITING;  
+        queue.insert(&previous->_link);
+
+        if(!Thread::_ready.empty()) {
+            Thread::_running = Thread::_ready.remove()->object();
+            Thread::_running->_state = Thread::RUNNING;
+            Thread::dispatch(previous, Thread::_running);
+        } else {
+            Thread::idle();
+        }
+
+        end_atomic();
+    } 
+    void wakeup() { 
+        begin_atomic();
+
+        if(!queue.empty()) {
+            wakeupThread();
+        }
+
+        end_atomic();
+
+        if(Thread::preemptive)
+            Thread::reschedule();
+    }
+    void wakeup_all() { 
+        begin_atomic();
+
+        while(!queue.empty()) {
+            wakeupThread();
+        }
+
+        end_atomic();
+
+        if(Thread::preemptive)
+            Thread::reschedule();
+    }
+
+    void wakeupThread(){
+        Thread* syncThread = queue.remove()->object();
+        syncThread->_state = Thread::READY;
+        Thread::_ready.insert(&syncThread->_link);
+    }
+
+private:
+    Thread::Queue queue;
 };
 
 __END_SYS
diff --git a/src/abstraction/semaphore.cc b/src/abstraction/semaphore.cc
index a2e8f7a..029df86 100644
--- a/src/abstraction/semaphore.cc
+++ b/src/abstraction/semaphore.cc
@@ -20,9 +20,11 @@ void Semaphore::p()
 {
     db<Synchronizer>(TRC) << "Semaphore::p(this=" << this << ",value=" << _value << ")" << endl;
 
-    fdec(_value);
-    while(_value < 0)
+    begin_atomic();
+    if(fdec(_value) < 1)
         sleep();
+    else
+        end_atomic();
 }
 
 
@@ -30,9 +32,11 @@ void Semaphore::v()
 {
     db<Synchronizer>(TRC) << "Semaphore::v(this=" << this << ",value=" << _value << ")" << endl;
 
-    finc(_value);
-    if(_value < 1)
+    begin_atomic();
+    if(finc(_value) < 0)
         wakeup();
+    else
+        end_atomic();
 }
 
 __END_SYS
diff --git a/src/abstraction/semaphore_test.cc b/src/abstraction/semaphore_test.cc
index e4033f4..9d4180e 100644
--- a/src/abstraction/semaphore_test.cc
+++ b/src/abstraction/semaphore_test.cc
@@ -11,7 +11,7 @@ using namespace EPOS;
 
 const int iterations = 10;
 
-Mutex table;
+Semaphore table;
 
 Thread * phil[5];
 Semaphore * chopstick[5];
@@ -25,20 +25,20 @@ int philosopher(int n, int l, int c)
 
     for(int i = iterations; i > 0; i--) {
 
-        table.lock();
+        table.p();
         Display::position(l, c);
         cout << "thinking";
-        table.unlock();
+        table.v();
 
         Delay thinking(2000000);
 
         chopstick[first]->p();   // get first chopstick
         chopstick[second]->p();   // get second chopstick
 
-        table.lock();
+        table.p();
         Display::position(l, c);
         cout << " eating ";
-        table.unlock();
+        table.v();
 
         Delay eating(1000000);
 
@@ -46,17 +46,17 @@ int philosopher(int n, int l, int c)
         chopstick[second]->v();   // release second chopstick
     }
 
-    table.lock();
+    table.p();
     Display::position(l, c);
     cout << "  done  ";
-    table.unlock();
+    table.v();
 
     return iterations;
 }
 
 int main()
 {
-    table.lock();
+    table.p();
     Display::clear();
     Display::position(0, 0);
     cout << "The Philosopher's Dinner:" << endl;
@@ -85,14 +85,15 @@ int main()
     Display::position(19, 0);
 
     cout << "The dinner is served ..." << endl;
-    table.unlock();
+    table.v();
 
     for(int i = 0; i < 5; i++) {
+        Delay init(1000000);
         int ret = phil[i]->join();
-        table.lock();
+        table.p();
         Display::position(20 + i, 0);
         cout << "Philosopher " << i << " ate " << ret << " times " << endl;
-        table.unlock();
+        table.v();
     }
 
     for(int i = 0; i < 5; i++)
